javatests/arcs/core/data/expression/ExpressionTest.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.expression
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+import kotlin.test.assertFailsWith
+
+/** Tests for [TypeVariable]. */
+@RunWith(JUnit4::class)
+class ExpressionTest {
+    fun <T> evalBool(expression: Expression<T>) = evalExpression<T, Boolean>(expression)
+    fun <T> evalNum(expression: Expression<T>) = evalExpression<T, Number>(expression)
+
+    @Test
+    fun testEvaluator() {
+        // numeric binary ops
+        assertThat(evalNum(2.asExpr() + 1.asExpr())).isEqualTo(3)
+        assertThat(evalNum(2.asExpr() - 1.asExpr())).isEqualTo(1)
+        assertThat(evalNum(2.asExpr() * 2.asExpr())).isEqualTo(4)
+        assertThat(evalNum(6.asExpr() / 3.asExpr())).isEqualTo(2)
+
+        // field ops
+        assertThat(evalNum<Number>(mapOf("foo" to 42).asScope()["foo"])).isEqualTo(42)
+
+        // query ops
+        assertThat(evalExpression<Number, Number>(query("arg"), "arg" to 42)).isEqualTo(42)
+
+        // Boolean ops
+        assertThat(evalBool(1.asExpr() lt 2.asExpr())).isTrue()
+        assertThat(evalBool(2.asExpr() lte 2.asExpr())).isTrue()
+        assertThat(evalBool(2.asExpr() gt 1.asExpr())).isTrue()
+        assertThat(evalBool(2.asExpr() gte 2.asExpr())).isTrue()
+        assertThat(evalBool((1.asExpr() lt 2.asExpr()) and (2.asExpr() gt 1.asExpr()))).isTrue()
+
+        // Unary ops
+        assertThat(evalNum(-2.asExpr())).isEqualTo(-2)
+        assertThat(evalBool(!(2.asExpr() lt 1.asExpr()))).isTrue()
+
+        // Equality ops
+        assertThat(evalBool(2.asExpr() eq 2.asExpr())).isTrue()
```

Comment:
some expressions that evaluate to false would be good for completeness, around here, e.g.

```
assertThat(evalBool(2.asExpr() lt 1.asExpr())).isFalse()
(...)
assertThat(evalBool(2.asExpr() eq 5.asExpr())).isFalse()
(...)
assertThat(evalBool("Hello".asExpr() neq "Hello".asExpr())).isFalse()
```